src/history/hs_rec.c

```diff
@@ -669,6 +669,26 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
                 F_SET(first_non_ts_upd, WT_UPDATE_CLEARED_HS);
             }
 
+            /* We should never write a prepared update to the history store. */
+            WT_ASSERT(session,
+              upd->prepare_state != WT_PREPARE_INPROGRESS &&
+                upd->prepare_state != WT_PREPARE_LOCKED);
+
+            /*
+             * Skip the check if we don't have a snapshot. Therefore, the check doesn't ensure that
+             * we never write uncommitted updates to the history store. But this is the best we can
+             * do for now.
+             *
+             * Sometimes the application threads and the checkpoint thread will fall behind the
+             * eviction threads, it may choose an invisible update to write to the disk if a failed
+             * eviction decided to write that update to the disk before. In this case, no point to
+             * check the visibility of the history store updates.
+             */
+            WT_ASSERT(session,
```

Comment:
We normally use `conn->txn_global->last_running` to check visibility of updates when there are no snapshots. We can use the same test here.

Reply:
No, we can't do that here. Some of the invisible updates will be selected because of the special case I described above.

Reply:
I am not sure how useful this check is here then. By the time eviction threads reach here, they would have already released the snapshot. Not sure though, but is it possible to move these checks up somewhere we have already the snapshot for eviction threads? I believe most of the reconcile work is either performed by checkpoint or eviction threads so we kind of miss almost half of the scenarios.